### PR TITLE
Bug/poisson sample runs forever

### PR DESCRIPTION
**Pull request by christophernorth**
Updated the sample function so it can deal with the numerical stability issues inherent with a large mu.  All existing tests pass and I've added a new one to ensure that large mu values stop correctly.

I haven't updated the mainline test as the sample function is O(mu) in complexity and thus large mus take a reasonable amount of time to sample from.

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java [hunk 1]
```diff
@@ -41,6 +36,13 @@ public void samplingProducesRealisticMeanAndStandardDeviation() {
         assertEquals(sd, standardDeviation, epsilon);
     }
 
+    @Test
+    public void largeMuIsSupported() {
```
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java]: Good Point - I'll put in a comment to cover this.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java [hunk 2]
```diff
@@ -41,6 +36,13 @@ public void samplingProducesRealisticMeanAndStandardDeviation() {
...
+        final double mu = 900.0;
+        PoissonVertex testVertex = new PoissonVertex(mu);
+
+        int sample = testVertex.sample().scalar();
```
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java]: Yep - will do.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java [hunk 3]
```diff
@@ -46,15 +46,29 @@ private static int sample(double mu, KeanuRandom random) {
...
-        }
+        final double STEP = 500;
+        double muLeft = mu;
+        int k = 0;
+        double p = 1.0;
+
+        do {
+            k++;
+            double u;
+            while ((u = random.nextDouble()) == 0);
```
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java]: Yep - I need a value in the interval (0, 1).  I can certainly wrap that in a new function though.

As for the value - I think this is a matter of taste, so I'm happy to go with what ever other people prefer.  Java will "promote" int 0 to double +0.0 when doing the comparison so the correct semantics are maintained, but if people prefer the 0.0 style, I'm happy with that (of course, there's a point that neither will actually match -0.0 but that's perhaps a point for another day :)

File: keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java [hunk 4]
```diff
@@ -23,15 +21,12 @@ public void samplingProducesRealisticMeanAndStandardDeviation() {
         KeanuRandom random = new KeanuRandom(1);
         PoissonVertex testPoissonVertex = new PoissonVertex(mu);
 
-        List<Integer> samples = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/intgr/probabilistic/PoissonVertexTest.java]: 👍

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java]: Yes, I wasn't suggesting it was incorrect logic to promote int to double, it just makes me feel icky.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/KeanuRandom.java [hunk 5]
```diff
@@ -70,6 +71,14 @@ public double nextDouble(double min, double max) {
         return nd4jRandom.nextDouble() * (max - min) + min;
     }
 
+    public double nextDoubleNonZero() {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/KeanuRandom.java]: Is this a general RNG type of function or something specific to Poisson?

Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/KeanuRandom.java]: I could see this being useful in other places (in fact, I'm planning on having a look through as I suspect we potentially have a few bugs hiding where we need numbers in (0, 1) but are actually getting them in [0, 1)

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java [hunk 6]
```diff
@@ -46,15 +46,29 @@ private static int sample(double mu, KeanuRandom random) {
...
+                    muLeft -= STEP;
+                } else {
+                    p *= Math.exp(muLeft);
+                    muLeft = 0;
+                }
+            }
+        } while (p > 1);
 
-        return i - 1;
+        return k - 1;
```
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Poisson.java]: Done

**Pull request merged**